torch/csrc/jit/codegen/cuda/lower_alias_memory.cpp

```diff
@@ -61,127 +63,903 @@ class SymbolicSizePrinter : private kir::IrVisitor {
+        TORCH_WARN_ONCE(
+            "Lower_alias_memory : dynamic sized register allocation");
+        return;
+      }
+      if (register_size.value() <= kRegisterSizeThreshold) {
+        should_try_alias = false;
       }
     }
+
+    auto data_type = kir_tv->dtype();
+    auto size_print = SymbolicSizePrinter::printSize(alloc);
+
+    // Make sure we don't have conflicting information on record
+    TORCH_INTERNAL_ASSERT(!map_allocate_to_info_.count(alloc));
+    TORCH_INTERNAL_ASSERT(!map_tv_to_allocations_.count(kir_tv->name()));
+
+    // make AllocationUseDefInfo:
+    auto alloc_info = makeUseDefInfo();
+    alloc_info->alloc_expr = alloc;
+    alloc_info->mem_type = mem_type;
+    alloc_info->data_type = data_type;
+    alloc_info->size_expr = size_print;
+    alloc_info->loop_info = current_stack_.back();
+    alloc_info->should_try_alias = should_try_alias;
+
+    // record short cuts
+    map_allocate_to_info_[alloc] = alloc_info;
+    map_tv_to_allocations_[kir_tv->name()] = alloc_info;
   }
 
- private:
-  // Do we have a true pointwise op?
-  // (ie. a TV op, excluding direct assignments and reductions)
-  static bool isPointwiseTvOp(const kir::Expr* expr) {
-    if (ir_utils::isTVOp(expr)) {
-      if (auto unary_op = dynamic_cast<const kir::UnaryOp*>(expr)) {
-        return unary_op->operation() != UnaryOpType::Set;
-      } else {
-        return expr->isA<kir::BinaryOp>() || expr->isA<kir::TernaryOp>();
+  void collectScopeUseDefInfo(const std::vector<kir::Expr*>& exprs) {
+    // Reset position pointer
+    resetExprCounter();
+    TORCH_INTERNAL_ASSERT(global_scope_info_ != nullptr);
+    current_stack_.push_back(global_scope_info_);
+    handleScope(exprs);
+  }
+
+  void collectScopeInfo(const std::vector<kir::Expr*>& exprs) {
+    // Reset position pointer
+    resetExprCounter();
+    collectScopeInfoWithinLoop(exprs, nullptr);
+  }
+
+  void collectScopeInfoWithinLoop(
+      const std::vector<kir::Expr*>& exprs,
+      kir::ForLoop* current_loop) {
+    auto loop_info = makeScopeInfo(current_loop);
+    for (auto expr : exprs) {
+      current_pos_++;
+      if (auto for_loop = dynamic_cast<kir::ForLoop*>(expr)) {
+        collectScopeInfoWithinLoop(for_loop->body().exprs(), for_loop);
       }
     }
-    return false;
+    loop_info->end_pos = current_pos_ + 1;
   }
 
-  // Find an Input Allocate that is compatible with the Output Allocate
-  const kir::Allocate* findCompatibleInputAllocate(
-      const DataType output_dtype,
-      const std::string& output_size_str,
-      const kir::Expr* expr) {
-    // Stop searching if current op is not point-wise
-    if (!isPointwiseTvOp(expr)) {
-      return nullptr;
+  void resetExprCounter() {
+    current_pos_ = -1;
+  }
+
+  //! Checks that the current loop nest is not realizing a serial
+  //!  broadcast so that each index of producer buffer will only
+  //!  be visited once.
+  bool indexEachLocalPointOnce(TensorView* producer, TensorView* consumer) {
```

Comment:
I'm not sure if mapping the id's with loop maps directly would be enough in this case. 

1. If I understand correctly, the `producer_id`'s active in allocation would map easily to the loops writing into the buffer, because we use the CA's when inserting the allocates. I didn't want to assume the loops reading from it will have the same or even similar loop structure, especially in the case of persistent buffers. Broadcast resolutions are usually read ops from the buffer. 

2. I wanted to be careful with concrete axes that have broadcast axes merged in so needed to check further than `producer_id->isBroadcast()`.

But I do think the current approach could be simplified. The current check in this PR is trying to be on the conservative side. The idea is basically:

for each expr reading from the buffer, 
  - inspect the loop structure this expr is in by grabbing all the `IterDomain`'s corresponding to the loop nest and putting all the root `IterDomain`'s used in deriving this loop structure into a set.
  - The check passes only if all the roots in the set from the previous step map to:
     - A concrete root domain of the producer or
     - A new broadcast axis if the expression is a broadcast op (used consumer to facilitate checking this)

Reply:
I think we talked about this, did we think index map would be sufficient here?